### ISSUE TITLE
JSON property is missing from schema when using conditional serialization

### ISSUE THREAD

**Issue by omfgicbf** [RSuter/NSwag#1992]
With a model that has a property that is hidden using ShouldSerialize, e.g:-

```
public class TestModel
{
    public bool Foo { get; set; }
    
    public bool Bar { get; set; }
    
    public bool ShouldSerializeFoo() => false;
}
```

The schema generated for this model excludes the `Foo` property when the model is used as a response (as expected) but it is also excluded when the module is used for a request.

As this `Foo` property is only excluded from serialization, and not from *de*serialization, I would expect it to be present in the schema.

In addition, if the conditional serialization of a property relies on a value that is set only under certain conditions at runtime (and which is not satisifed when NSwag is run), the property is exlcuded from the spec.

**Comment by RSuter** [comment]
Conditional serialization is hard to express in a schema as the generator does not know whether the type is used as input or output...

**Comment by omfgicbf** [comment]
Isn't the determination on whether the model is being used for input or output made based on where it appears? i.e. if it's a return value, then it's output; if it's an argument to a controller method, then input?

Aren't both those scenarios handled by different processors?

**Comment by RSuter** [comment]
There is no difference between inputs and outputs - this is also more complex as you would also need to correctly track child schemas... additionally the Newtonsoft.Json contract resolver is not really exposing enough information I think. So implementing this would probably be a huge undertaking - also when keeping the new .net core JSON serializers in mind...

**Comment by omfgicbf** [comment]
Right; so something like a serialize/deserialize flag on ITypeMapper so cached models can be split between serialized and deserialized versions and referenced accordingly (which could default to deserialize as it does now), and a custom contract resolver?

**Comment by RSuter** [comment]
Isnt this discussion about readonly and writeonly properties? ie your initial sample should set readonly on foo to false?

**Comment by omfgicbf** [comment]
It's not necessarily about read-only/write-only; it's about the parameter appearing at all (though they amount to the same thing I guess).

My thinking is that it doesn't make sense for a write-only property to appear in a response model (and vice-versa). However, if it is at least marked as read-only or write-only, the it can just be omitted during rendering.

Anyway, that was a bit of a side-track; the issue here is that I've designated `Foo` as write-only (using `ShouldSerialize`), and the property doesn't appear at all in the model.

It appears `ShouldSerializeFoo` is evaluated rather than just reflected; if it happens to evaluate to false at the time it's inspected, it's not included in the model.

**Comment by icnocop** [comment]
Related: https://github.com/RicoSuter/NSwag/issues/4306

**Comment by kdrapel** [comment]
My grain of salt, I had to use the following syntax to grab the serializer settings, this may be helpful for someone:

```
            services.AddOpenApiDocument((settings, serviceProvider) =>
            {

                NJsonSchema.NewtonsoftJson.Generation.NewtonsoftJsonSchemaGeneratorSettings jsonSettings = (NJsonSchema.NewtonsoftJson.Generation.NewtonsoftJsonSchemaGeneratorSettings) settings.SchemaSettings; 
```
